### PR TITLE
Draft for catch2 testing framework usage

### PR DESCRIPTION
**Pull request by senior-zero**
Cub uses testing facilities that are far from perfect. This PR contains Catch2 integration and a few convenience wrappers. Main advantages of Catch2:

- Readable way of specifying cartesian products of types/compile-time parameters. Instead of having
  ```cuda
  template <
      typename                    SampleT,
      int                         BINS,
      int                         BLOCK_THREADS,
      int                         ITEMS_PER_THREAD>
  void Test()
  {
      Test<SampleT, BINS, BLOCK_THREADS, ITEMS_PER_THREAD, BLOCK_HISTO_SORT>();
      Test<SampleT, BINS, BLOCK_THREADS, ITEMS_PER_THREAD, BLOCK_HISTO_ATOMIC>();
  }
  
  template <
      typename                    SampleT,
      int                         BINS,
      int                         BLOCK_THREADS>
  void Test()
  {
      Test<SampleT, BINS, BLOCK_THREADS, 1>();
      Test<SampleT, BINS, BLOCK_THREADS, 5>();
  }
  

  template <
      typename                    SampleT,
      int                         BINS>
  void Test()
  {
      Test<SampleT, BINS, 32>();
      Test<SampleT, BINS, 96>();
      Test<SampleT, BINS, 128>();
  }
  ```
  Catch2 allows writing:
  ```cuda
  using types = catch2_helper::type_list<unsigned char, unsigned short>;
  using threads_in_block = catch2_helper::enum_type_list<int, 32, 256, 1024>;
  using items_per_thread = catch2_helper::enum_type_list<int, 1, 5>;
  using bins = catch2_helper::enum_type_list<int, 32, 96, 128>;
  using algorithms = catch2_helper::enum_type_list<BlockHistogramAlgorithm,
                                                   BLOCK_HISTO_SORT,
                                                   BLOCK_HISTO_ATOMIC>;
  
  using cases = catch2_helper::cartesian_product<
    catch2_helper::type_list<
      types,
      items_per_thread,
      threads_in_block,
      bins,
      algorithms>>;
  
  TEMPLATE_LIST_TEST_CASE("Block histogram can be computed with uniform input",
                          "[histogram][block]",
                          cases)
  ```
- Readable generators. Instead of traversing parameter space in multiple nested loops:
  ```cuda
  void Test(thrust::default_random_engine &rng)
  {
    for (unsigned int num_items = ItemsPerThread * ThreadsInBlock;
         num_items > 1;
         num_items /= 2)
    {
  ```
  Catch2 provides readable generator expressions:
  ```cuda
    const int num_keys = GENERATE_COPY(take(10, random(0, ItemsPerThread * ThreadsInBlock)));
  ```
- Display of variables' content in case of failure. Currently, only the place of failure is displayed. Catch2 displays the content of compared variables:
  ```cuda
  thrust::host_vector<KeyT> h_reference = /* ... */;
  REQUIRE( h_reference == d_keys );
  ```
  ```
  ctest_block_merge_sort.cu(180): FAILED:
  REQUIRE( h_reference == d_keys ) with expansion:
  { 120, -2147483648, -2147483648, -2147483648 }
  ==
  { -2147483648, -2147483648, -2147483648, -2147483648 }
  ```
- Breaking into the debugger. Catch2 automatically creates a breakpoint in case of failure. It's not the case with the current approach. 
- Ability to execute only part of unit-tests in a file. It's also possible to perform tests by wildcard or tags. I aggregate all files that start with `ctest_` into a single test (per configuration). So it'll be possible to perform all the histogram tests (across scopes) or all of the block-scope tests depending on the command line parameters. 

It's not the full list of features. I'd like to discuss this draft in case there are any corner cases I'm missing. If there's a better unit-testing framework I'd also like to discuss it here.

### CODE REVIEW

**Comment by jrhemstad** [comment]
Huge +1 from me. I experimented with using Catch2 in [cuCollections ](https://github.com/NVIDIA/cuCollections/tree/dev/tests) and I have loved it. 

I know CUB doesn't currently use GTest, but many of my favorite features are in comparison to what you can do with GTest:

- Compared to GTest, it cuts out a lot of the boiler plate of writing test fixture clases. The independent, depth-first execution of nested of `SECTION`s makes it natural and intuitive to reuse common code but maintain independent execution paths. 
- This may seem silly, but I also really love that you can include plain text description of what the intent of a particular test/section is, e.g.:
```
TEST_CASE( "vectors can be sized and resized", "[vector]" ) {
    std::vector<int> v( 5 );

    REQUIRE( v.size() == 5 );
    REQUIRE( v.capacity() >= 5 );

    SECTION( "resizing bigger changes size and capacity" ) {
        v.resize( 10 );

        REQUIRE( v.size() == 10 );
        REQUIRE( v.capacity() >= 10 );
    }
}
```
- Another big advantage compared to Gtest is you can use `__device__` lambdas in the body of a catch2 test. With GTest where a test section is expanded into a private member function of a class. nvcc forbids a `__device__` lambda in such a function.

**Comment by senior-zero** [comment]
> Huge +1 from me. I experimented with using Catch2 in [cuCollections ](https://github.com/NVIDIA/cuCollections/tree/dev/tests) and I have loved it.
> 
> I know CUB doesn't currently use GTest, but many of my favorite features are in comparison to what you can do with GTest:
> 
> * Compared to GTest, it cuts out a lot of the boilerplate of writing test fixture clases. The independent, depth-first execution of nested of `SECTION`s makes it natural and intuitive to reuse common code but maintain independent execution paths.
> * This may seem silly, but I also really love that you can include plain text description of what the intent of a particular test/section is, e.g.:
> 
> ```
> TEST_CASE( "vectors can be sized and resized", "[vector]" ) {
>     std::vector<int> v( 5 );
> 
>     REQUIRE( v.size() == 5 );
>     REQUIRE( v.capacity() >= 5 );
> 
>     SECTION( "resizing bigger changes size and capacity" ) {
>         v.resize( 10 );
> 
>         REQUIRE( v.size() == 10 );
>         REQUIRE( v.capacity() >= 10 );
>     }
> }
> ```
> 
> * Another big advantage compared to Gtest is you can use `__device__` lambdas in the body of a catch2 test. With GTest where a test section is expanded into a private member function of a class. nvcc forbids a `__device__` lambda in such a function.

Initially, I've integrated GTest. Both generators and sections convinced me that it's better to use Catch2 😄 Have you encountered an issue related to Catch2 usage in `.cu` files? Currently I have to suppress the following warning to get it working:

```cuda
#if CUB_HOST_COMPILER == CUB_HOST_COMPILER_MSVC
#pragma diag_suppress 177
#endif
```

**Comment by jrhemstad** [comment]
> Have you encountered an issue related to Catch2 usage in `.cu` files?

All the test files in cuCollections are `.cu` files: https://github.com/NVIDIA/cuCollections/tree/dev/tests

The is one warning that's generated when using a `__device__` lambda in a test due to the fact that a Catch2 test expands out to a function in an anonymous namespace and the capture generated by `nvcc` for `__device__` lambda takes the address of the enclosing function and uses it as a template parameter. So it complains about using the address of an anonymous function as part of the type signature. 

What warning are you seeing?

**Comment by senior-zero** [comment]
> > Have you encountered an issue related to Catch2 usage in `.cu` files?
> 
> All the test files in cuCollections are `.cu` files: https://github.com/NVIDIA/cuCollections/tree/dev/tests
> 
> The is one warning that's generated when using a `__device__` lambda in a test due to the fact that a Catch2 test expands out to a function in an anonymous namespace and the capture generated by `nvcc` for `__device__` lambda takes the address of the enclosing function and uses it as a template parameter. So it complains about using the address of an anonymous function as part of the type signature.
> 
> What warning are you seeing?

I get a bunch of unused variables warnings:
```
catch.hpp(15860): error #177-D: variable "colour" was declared but never referenced
```
I get it with MSVC 16.11.1 + C++17. This issue appears only in `.cu` files.

File: test/catch_runner.cu [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+int DeviceGuard(int device_id)
+{
+  int device_count {};
+  if (cudaGetDeviceCount(&device_count) != cudaSuccess)
+  {
+    throw std::runtime_error("Can't query devices number");
+  }
+
+  if (device_id >= device_count)
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/catch_runner.cu:47]: I've duplicated the old logic:

```cuda
DeviceInit:
if ((dev > deviceCount - 1) || (dev < 0))
{
  dev = 0;
}
```

I agree, it's better to exit with an error.

**Comment by senior-zero** [comment]
gpuCI: NVIDIA/thrust#1559

File: test/catch2_test_helper.h [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+#if CUB_HOST_COMPILER == CUB_HOST_COMPILER_MSVC
+#pragma diag_suppress 177
+#elif CUB_HOST_COMPILER == CUB_HOST_COMPILER_CLANG
+#pragma clang diagnostic ignored "-Wunused-variable"
+#endif
+
+#include <catch2/catch.hpp>
+
+// It's a port of cartesian product from https://github.com/NVIDIA/nvbench
+namespace catch2_helper
```
Comment (jrhemstad) [inline test/catch2_test_helper.h:51]: I'd love if we could contribute this to Catch2. I filed an issue for exactly this a long time ago https://github.com/catchorg/Catch2/issues/1989

File: CMakeLists.txt [hunk 3]
```diff
@@ -117,6 +117,9 @@ if (CUB_ENABLE_TESTING OR CUB_ENABLE_EXAMPLES)
 endif()
 
 if (CUB_ENABLE_TESTING)
+  include(cmake/CPM.cmake)
+  CPMAddPackage("gh:catchorg/Catch2@2.13.7")
```
Comment (jrhemstad) [inline CMakeLists.txt:121]: Since we're adding this fresh, we may want to look at using v3 right away. In cuCo I found the compile time can be quite bad, and that's precisely what v3 is meant to address. 

https://github.com/catchorg/Catch2/blob/devel/docs/migrate-v2-to-v3.md#top

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -78,9 +78,15 @@ file(GLOB test_srcs
   test_*.cu
 )
 
+file(GLOB catch2_test_srcs
+  RELATIVE ${CUB_SOURCE_DIR}/test
+  CONFIGURE_DEPENDS
+  catch2_test_*.cu
```
Comment (miscco) [inline test/CMakeLists.txt:84]: Do we really need the `catch2_` prefix?

Comment (senior-zero) [reply test/CMakeLists.txt:84]: After everything is settled - I don't think we need one. Right now we have `test_` prefix for existing infrastructure, which is going to coexist for a while.

Comment (miscco) [reply test/CMakeLists.txt:84]: Yeah sorry for the noise, I should have checked that we have more than 3 test files ;)

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -222,4 +228,164 @@ foreach (test_src IN LISTS test_srcs)
...
+    target_compile_options(${target} PRIVATE "-fPIC")
+  endif()
+
+  target_include_directories(${target} 
+    PUBLIC "${CUB_SOURCE_DIR}/test")
+
+  cub_clone_target_properties(${target} ${cub_target})
+endforeach()
+
+function(cub_add_catch2_test src name defs target config_prefix)
```
Comment (allisonvacanti) [inline test/CMakeLists.txt:263]: Add a docstring similar to `cub_add_test`. We should try to make the signatures match if possible.

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -222,4 +228,164 @@ foreach (test_src IN LISTS test_srcs)
...
+        message(VERBOSE "  ${i}: ${test_name} ${label} ${defs}")
+      endforeach()
+    endif()
+
+    cub_get_target_property(config_prefix ${cub_target} PREFIX)
+
+    if (num_variants EQUAL 0)
+      # Only one version of this test.
+      cub_add_catch2_test(
+        ${test_src}
```
Comment (allisonvacanti) [inline test/CMakeLists.txt:322]:
```suggestion
        "${test_src}"
```

File: test/CMakeLists.txt [hunk 7]
```diff
@@ -222,4 +228,164 @@ foreach (test_src IN LISTS test_srcs)
...
+  # Add to the active configuration's meta target
+  add_dependencies(${config_meta_target} ${test_target})
+
+  # Meta target that builds tests with this name for all configurations:
+  if (NOT TARGET ${test_meta_target})
+    add_custom_target(${test_meta_target})
+  endif()
+  add_dependencies(${test_meta_target} ${test_target})
+
+  add_test(NAME ${test_target} COMMAND "$<TARGET_FILE:${test_target}>")
```
[the remark this answers is not in the document]
Comment (allisonvacanti) [reply test/CMakeLists.txt:388]: Might be worth seeing if all of this "setup metatargets, link ${cub_target}, copy props, register CTest, etc" logic could be abstracted out into a `cub_setup_test(test_target cub_target)` function and reused.

There are three copies of this logic currently in this file, but I haven't looked too closely at whether they overlap enough to share an impl.

**Review by allisonvacanti — CHANGES_REQUESTED** [review]
Only reviewed the CMake logic. Really excited to see this progress :)

File: test/CMakeLists.txt [hunk 8]
```diff
@@ -222,4 +228,164 @@ foreach (test_src IN LISTS test_srcs)
...
+      endforeach()
+    endif()
+  endforeach()
+
+  target_link_libraries(${test_target} PRIVATE
+          ${cub_target}
+          Catch2::Catch2)
+
+  if ("NVCXX" STREQUAL "${CMAKE_CUDA_COMPILER_ID}")
+    target_link_options(${test_target} PRIVATE "-cuda")
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/CMakeLists.txt:373]: Because we never linked cuda libraries

[the remark this answers is not in the document]
Comment (allisonvacanti) [reply test/CMakeLists.txt:374]: Gotcha.

File: test/CMakeLists.txt [hunk 9]
```diff
@@ -222,4 +228,164 @@ foreach (test_src IN LISTS test_srcs)
...
+    catch2_test_helper.h 
+    c2h/custom_type.cuh 
+    c2h/generators.cuh 
+    c2h/generators.cu)
+
+  target_link_libraries(${target} PRIVATE CUDA::curand)
+
+  set_property(TARGET ${target} PROPERTY POSITION_INDEPENDENT_CODE ON)
+
+  target_link_options(${target} PRIVATE "-shared")
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/CMakeLists.txt:259]: Without this nvhpc fails:
```bash
:cmake --version                            
cmake version 3.22.1
:nvc++ --version                   
nvc++ 22.5-0 64-bit target on x86-64 Linux -tp haswell
: && /nvc++ -cuda -Wl,-soname,libcub.cpp17.test.catch2_helper.so -o lib/libcub.cpp17.test.catch2_helper.so test/CMakeFiles/cub.cpp17.test.catch2_helper.dir/c2h/generators.cu.o  -Wl,-rpath,/usr/local/cuda/lib64  /usr/local/cuda/lib64/libcurand.so && :
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/crt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
```
I'll add this option only for the nvhpc case.

File: docs/test_overview.md [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+using type = c2h::custom_type_t<c2h::accumulateable_t,
+                                c2h::equal_comparable_t>;
+```
+
+Here we enumerate all the type properties that we are interested in.
+The produced type ends up having `operator==` and `operator+`.
+There are more properties implemented. 
+If some property is missing, 
+it'd be better to add one in `c2h` 
+instead of writing a custom type from scratch.
```
Comment (elstehle) [inline docs/test_overview.md:120]: Wondering if we should give this part its own section or fuse it with _Type Lists_

**Review by elstehle — COMMENTED** [review]
Thank you for adding the test overview, I find it immensely helpful. 

_This is a partial review, limited to `test_overview.md`_

**Review by elstehle — COMMENTED** [review]
Flushing a few more comments. Looks great so far! This partial review includes:
- `c2h/*`
- `test/catch2_test_block_run_length_decode.cu`
- `test/catch2_test_block_merge_sort.cu`
- `test/catch2_test_block_store.cu`

Will finish these remaining parts today:
- `test/catch2_test_block_adjacent_difference.cu`
- `test/catch2_test_block_histogram.cu`

File: test/c2h/generators.cu [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+
+RANDOM_TO_VEC_ITEM_SPEC(0, x);
+RANDOM_TO_VEC_ITEM_SPEC(1, y);
+RANDOM_TO_VEC_ITEM_SPEC(2, z);
+RANDOM_TO_VEC_ITEM_SPEC(3, w);
+
+class generator_state_t
+{
+public:
+};
```
Comment (elstehle) [inline test/c2h/generators.cu:129]: It seems this isn't used

**Review by elstehle — COMMENTED** [review]
Great work! Really looking forward to having this in and start building on it!
Completing my review on these files with just few minor comments/questions:
- `c2h/*`
- `test/catch2_test_block_run_length_decode.cu`
- `test/catch2_test_block_merge_sort.cu`
- `test/catch2_test_block_store.cu`
- `test/catch2_test_block_adjacent_difference.cu`
- `test/catch2_test_block_histogram.cu`

File: test/c2h/custom_type.cuh [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+{
+  // The CUDA compiler follows the IA64 ABI for class layout, while the 
+  // Microsoft host compiler does not.
+  char workaround_msvc;
+
+public:
+  __host__ __device__ bool operator<(const CustomType& other) const
+  {
+    return static_cast<const CustomType&>(*this).get_key() 
+         < other.get_key();
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/c2h/custom_type.cuh:76]: This form of `less_comparable` is needed to check stability of our algorithms. If we were to compare value as well, we wouldn't be able to check that `{1, 2}` and `{1, 3}` are permuted. I think if we find a case where this is an issue, we will introduce something like `lexicographically_less_comparable_t` or `pair_less_comparable_t` property later.

File: test/catch2_test_block_run_length_decode.cu [hunk 13]
```diff
@@ -39,25 +33,25 @@
...
  * \tparam RunLengthsItT Iterator type providing the runs' lengths
  * \tparam RUNS_PER_THREAD The number of runs that each thread is getting assigned to
- * \tparam DECODED_ITEMS_PER_THREAD The number of run-length decoded items that each thread is decoding
- * \tparam TEST_RELATIVE_OFFSETS_ Whether to also retrieve each decoded item's relative offset within its run
- * \tparam TEST_RUN_OFFSETS_ Whether to pass in each run's offset instead of each run's length
- * \tparam BLOCK_DIM_X The thread block length in threads along the X dimension
+ * \tparam DECODED_ITEMS_PER_THREAD The number of run-length decoded items that each thread is
+ * decoding \tparam TEST_RELATIVE_OFFSETS_ Whether to also retrieve each decoded item's relative
+ * offset within its run \tparam TEST_RUN_OFFSETS_ Whether to pass in each run's offset instead of
+ * each run's length \tparam BLOCK_DIM_X The thread block length in threads along the X dimension
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/catch2_test_block_run_length_decode.cu:54]: Good point, thanks!

File: test/catch2_test_block_merge_sort.cu [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+        thrust::raw_pointer_cast(vals.data()),
+        static_cast<int>(keys.size()),
+        std::numeric_limits<T>::max(),
+        action);
+
+  REQUIRE( cudaSuccess == cudaPeekAtLastError() );
+  REQUIRE( cudaSuccess == cudaDeviceSynchronize() );
+}
+
+// %PARAM% THREADS_IN_BLOCK bs 64:256
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/catch2_test_block_merge_sort.cu:241]: Merge doesn't support non-power-of-two threads per block. In a presence of a warp merge sort, I don't think there's a huge value in testing 32 threads.

File: test/catch2_test_block_histogram.cu [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+  using sample_t = typename params::sample_t;
+
+  // Allocate problem device arrays
+  thrust::device_vector<int> d_histogram(params::bins);
+  thrust::device_vector<sample_t> d_samples(params::num_samples);
+
+  c2h::gen(CUB_SEED(10),
+           d_samples,
+           static_cast<sample_t>(0),
+           static_cast<sample_t>(params::bins - 1));
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/catch2_test_block_histogram.cu:200]: I think that the test is short enough (20 lines) not to bother with that :)

File: test/catch2_test_block_adjacent_difference.cu [hunk 16]
```diff
@@ -0,0 +1,423 @@
+/******************************************************************************
```
[the remark this answers is not in the document]
Comment (senior-zero) [reply test/catch2_test_block_adjacent_difference.cu:1]: I think it should be tested. Added TODO for now, thanks!

File: test/c2h/custom_type.cuh [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+  {
+    const CustomType& self = static_cast<const CustomType&>(*this);
+    
+    return self.get_key() == other.get_key() &&
+           self.get_val() == other.get_val();
+  }
+};
+
+template <class CustomType>
+class substractable_t
```
Comment (kaatish) [inline test/c2h/custom_type.cuh:98]:
```suggestion
class subtractable_t
```
typo

Comment (senior-zero) [reply test/c2h/custom_type.cuh:98]: It seems that `subtract` is used more often, thanks!

**Review by elstehle — APPROVED** [review]

**Review by kaatish — APPROVED** [review]

**Review by allisonvacanti — APPROVED** [review]
